config/Dockerfiles/fedoraci-runner/package-test/package-test.sh

```diff
@@ -175,6 +176,34 @@ if [[ `ls -1 tests*.yml 2>/dev/null | wc -l` == 0 ]]; then
     exit 1
 fi
 
+# Merge two results.yml files.
+#
+# This function merges previous results.yml into current results.yml.
+# If the two results.yml files are the same, it does nothing.
+#
+# Arguments:
+# - path to the previous results.yml file (e.g.: ${TEST_ARTIFACTS}/results.yml.prev)
+# - path to the current results.yml file
+function merge_results {
+
+    local old_results_path=${1}
+    local new_results_path=${2}
+
+    if ! [ -f "${old_results_path}" ] || ! [ -f "${new_results_path}" ]; then
+        # nothing to merge if the files don't exist
+        return
+    fi
+
+    local diff_rc=$( diff -q "${old_results_path}" "${new_results_path}" > /dev/null 2>&1 ; echo $? )
+    if [ "${diff_rc}" -eq 0 ]; then
+        # the files are the same
+        return
+    fi
```

Comment:
There's no need for the ${diff_rc} variable here.  You can write the above as:

diff -q "${old_results_path}" "${new_results_path}" > /dev/null 2>&1
[ $? -eq 0 ] && return

Reply:
Ah, right. Good catch. I will fix this.

Reply:
Oh, wait. Now I remember why I introduced that `diff_rc` variable.

There is `set -xeo pipefail` at the beginning of the script so it would bail out on that non-zero status code on this line.

Reply:
Right, but there's no pipe with the diff -q command.  It's just running diff to check if there are differences or not.  That's just checking the exit code.  If you want it to fail in the event that diff doesn't exist, I would add a check at the top of the script for required tools.  For diff, I would do:

PATH=/bin:/usr/bin

diff --help >/dev/null 2>&1
if [ $? -eq 127 ]; then
    echo "*** diff not found in PATH, exiting" >&2
    exit 1
fi

Reply:
I think it's the `e` in `set -xeo pipefail` that would cause trouble here. 

From man page:

```
-e      Exit immediately if a pipeline (which may consist of a single simple command), a list, or a  compound  command  (see  SHELL  GRAMMAR
        above),  exits  with  a  non-zero status.
        ...
```

`diff` returns 1 when the files are not the same, so the whole script would fail right there. And we actually want to continue and merge the files if they are not the same :)

Reply:
I guess I could do:
`diff -q "${old_results_path}" "${new_results_path}" > /dev/null 2>&1 && return`

WDYT?

Reply:
That's true about pipefail.  I prefer not to use it if possible because it's specific to bash here and it also makes debugging difficult.  In many cases you can take corrective action in the script based on the return code.  In the case of diff(1), it returns 0 if the files match, 1 if they are different, and 2 if there was a problem reading them.  The way it's written is matching causes an early exit.  There's no special handling for return code 2 and since most people don't memorize what return code 2 from diff means (I just had to look at the source), a pipefail error still won't help with debugging and maintenance later.  After further evaluation, I would do:

```
diff -q "${old_results_path}" "${new_results_path}" >/dev/null 2>&1
ret=$?
if [ ${ret} -eq 0 ]; then
    return
elif [ ${ret} -eq 2 ]; then
    echo "*** diff(1) read error" >&2
    exit 1
fi
```

I'd also put the block at the top to ensure diff is available and running it does not return 127.

If you go the pipefail route, the script will be bash-specific and you won't get detailed error handling on recovery capability.

Reply:
Good catch with the "2" status code.

To be honest, I am kinda nervous about removing that `set -e` from the script completely. I am not the original author of the script so I don't know which parts of the script rely on it.

So I propose the following change:

```
set +e
diff -q "${old_results_path}" "${new_results_path}" >/dev/null 2>&1
ret=$?
set -e
if [ ${ret} -eq 0 ]; then
    return
elif [ ${ret} -eq 2 ]; then
    echo "*** diff(1) read error" >&2
    exit 1
fi
```

WDYT?

Reply:
That looks ok, but don't you mean "set -e" for the first line?  "set -e" is the "exit immediately on failure" option.  "set +e" is a bash-specific command that disables that.  If you want "set +e", then the portable way to do that is to just remove the line.

Reply:
Sorry, wait, I didn't catch the second "set -e".  Yes, this looks fine except I would just put "set -e" at the top entirely and remove the embedded instances.

Reply:
> I would just put "set -e" at the top entirely and remove the embedded instances.

It's already there :) At the beginning of the script:
`set -xeo pipefail`

However, that would cause the `diff -q` line to exit the script if the files are not the same. And we don't want to that.

I updated the pull request.